File: ChatGPT/OCaml/response.ml
(* OCaml code for a Sudoku solver *

(* Step 1: Read a 9x9 matrix from a file into a 1D array, ignoring comments *)
let read_matrix filename =
  let ch = open_in filename in
  let matrix = Array.make 81 0 in
  let rec read_line i =
    if i < 81 then
      try
        let line = input_line ch in
        if String.get line 0 <> '#' then begin
          let nums = List.map int_of_string (String.split_on_char ',' (String.trim line)) in
          List.iteri (fun j n -> matrix.(i + j) <- n) nums;
          read_line (i + 9)
        end else
          read_line i
      with End_of_file -> ()
    else
      ()
  in
  read_line 0;
  close_in ch;
  matrix

(* Step 2: Calculate the complexity of a 9x9 matrix *)
let calculate_complexity matrix =
  Array.fold_left (fun acc x -> if x = 0 then acc + 1 else acc) 0 matrix

(* Step 3: Print the board in a 9x9 grid and the complexity for the unsolved board *)
let print_board matrix =
  Printf.printf "Complexity: %d\n" (calculate_complexity matrix);
  Array.iteri (fun i x ->
    Printf.printf "%d ";
    if (i + 1) mod 9 = 0 then Printf.printf "\n") matrix

(* Step 4: Solve the Sudoku board using a backtracking algorithm *)
let solve_board matrix =
  let is_valid n i =
    let row = i / 9 and col = i mod 9 in
    let block_row = row / 3 * 3 and block_col = col / 3 * 3 in
    for j = 0 to 8 do
      let row_idx = 9 * row + j in
      let col_idx = col + 9 * j in
      let block_idx = block_row * 9 + block_col + (j / 3) * 9 + (j mod 3) in
      if matrix.(row_idx) = n || matrix.(col_idx) = n || matrix.(block_idx) = n then
        return false
    done;
    true
  in
  let rec backtrack i count =
    if i = 81 then
      (true, count)
    else if matrix.(i) <> 0 then
      backtrack (i + 1) count
    else
      let rec try_num n count =
        if n = 10 then
          (false, count)
        else if is_valid n i then
          begin
            matrix.(i) <- n;
            let (solved, count) = backtrack (i + 1) (count + 1) in
            if solved then
              (true, count)
            else
              (matrix.(i) <- 0; try_num (n + 1) count)
          end
        else
          try_num (n + 1) count
      in
      try_num 1 count
  in
  let (solved, count) = backtrack 0 0 in
  (solved, matrix, count)

(* Step 5: Print the final board and the number of iterations with commas for thousands *)
let print_solution (solved, matrix, count) =
  if solved then begin
    print_board matrix;
    Printf.printf "Solved with %s iterations\n" (String.concat "," (List.map string_of_int (List.rev (Utils.int_to_list count))))
  end else
    Printf.printf "No solution found\n"

(* Step 6: Read matrices from the command line and solve them *)
let () =
  let filename = Sys.argv.(1) in
  let matrix = read_matrix filename in
  print_board matrix;
  let solution = solve_board matrix in
  print_solution solution

(* Helper function to convert an integer to a list of its digits *)
module Utils = struct
  let rec int_to_list n =
    if n = 0 then [] else int_to_list (n / 10) @ [n mod 10]
end
)

File: ChatGPT/Typescript/response.ts
const fs = require('fs');

// Step 1: Read a 9x9 space-separated matrix from a file and return it as a 1D array.
function readSudokuBoardFromFile(filePath: string): number[] {
  const fileContent = fs.readFileSync(filePath, 'utf-8');
  let board: number[] = [];
  fileContent.split('\n').forEach((line) => {
    if (!line.startsWith('#')) {
      const row = line.split(' ').map(Number);
      board = [...board, ...row];
    }
  });
  return board;
}

// Step 2: Calculate the complexity of a 9x9 matrix.
function calculateComplexity(board: number[]): number {
  return board.reduce((acc, val) => acc + (val === 0 ? 1 : 0), 0);
}

// Step 3: Print the board in a 9x9 grid.
function printBoard(board: number[]): void {
  console.log('\nSudoku Board:');
  for (let i = 0; i < board.length; i++) {
    if (i % 9 === 0 && i !== 0) console.log('');
    process.stdout.write(`${board[i]} `);
  }
  console.log('\n');
}

// Step 4: Solve the Sudoku board using a backtracking algorithm.
function solveSudoku(board: number[]): boolean {
  const size = 9;
  const boxSize = 3;

  function findEmptyPosition(board: number[]): number[] | null {
    for (let i = 0; i < size * size; i++) {
      if (board[i] === 0) return [Math.floor(i / size), i % size];
    }
    return null;
  }

  function isValidPlacement(board: number[], num: number, pos: [number, number]): boolean {
    const [row, col] = pos;

    // Check rows and columns
    for (let i = 0; i < size; i++) {
      if (board[row * size + i] === num || board[i * size + col] === num) return false;
    }

    // Check box
    const boxRow = Math.floor(row / boxSize) * boxSize;
    const boxCol = Math.floor(col / boxSize) * boxSize;

    for (let r = 0; r < boxSize; r++) {
      for (let c = 0; c < boxSize; c++) {
        if (board[(boxRow + r) * size + (boxCol + c)] === num) return false;
      }
    }

    return true;
  }

  const emptyPos = findEmptyPosition(board);

  if (!emptyPos) return true;

  for (let num = 1; num <= size; num++) {
    if (isValidPlacement(board, num, emptyPos)) {
      const [row, col] = emptyPos;
      board[row * size + col] = num;
      iterations++;

      if (solveSudoku(board)) return true;

      board[row * size + col] = 0; // backtrack
    }
  }

  return false;
}

// Step 5: Print the final board and the number of iterations.
let iterations = 0;
function printSolution(board: number[]): void {
  printBoard(board);
  console.log(`Solved in ${iterations.toLocaleString()} iterations.`);
}

// Step 6: The matrices to read will be submitted on the command line.
const filePath = process.argv[2];
const sudokuBoard = readSudokuBoardFromFile(filePath);
printBoard(sudokuBoard);
console.log(`Complexity: ${calculateComplexity(sudokuBoard)}`);

if (solveSudoku(sudokuBoard)) {
  printSolution(sudokuBoard);
} else {
  console.log('No solution found.');
}
